common/docs/rpcUtil.1
.\"
.\" $Log: rpcUtil.1,v $
.\" Revision 1.3  1994/02/24 20:54:28  markc
.\" Extended man page.
.\"
# Revision 1.2  1994/02/24  07:56:34  markc
# Added more comments.
#
.\"
.TH RPCUTIL 1L PARADYN "\\n(mo/\\n(dy/19\\n(yr"
.SH NAME
rpcUtil \- Support Library
.SH DESCRIPTION
.PP
This library provides utility functions.  It provides a list class, a hash table class, and functions to support client-server communications using XDR/pvm/threads.
.SH "LIST CLASS SYNOPSIS"
.B template <class Type> class List {
.br
.B	public:
.br
.B	List();
.br
.B	int  empty();
.br
.B	friend ostream &operator<<(ostream&, List<Type>&);
.br
.B	void print();
.br
.B	void add(Type data, void *key);
.br
.B	void add(Type data);
.br
.B	Boolean addUnique(Type data);
.br
.B	Boolean addUnique(Type data, void *key);
.br
.B	Type find(void *key);
.br
.B	Boolean remove(void *key);
.br
.B	int count();
.br
.B	Type operator *();
.br
.B	void operator +=(List<Type> mergee);
.br
.B	Type operator ++();
.br
.B	void map (void (*map_function)(const Type item));
.br
.B	void setCurrent();
.br
.B	const Type getCurrent();
.br
.B	const Type next();
.br
.B	void advanceCurrent();
.br
.B	protected:
.br
.B	ListItem<Type>	*head;
.br
.B	ListItem<Type>  *current;
.br
};
.br
.SH "HASH TABLE CLASS SYNOPSIS"
.PP
.B template <class Type> class HTable {
.br
.B    public:
.br
.B	friend ostream &operator<<(ostream&, HTable<Type>&);
.br
.B	void print();
.br
.B	HTable(Type data);
.br
.B	void add(Type data, void *key);
.br
.B	HTable(); 
.br
.B	Boolean addUnique(Type data, void *key);
.br
.B	Type find(void *key);
.br
.B	Boolean remove(void *key);
.br
.B	Type operator =(HTable<Type> arg);
.br
.B	Type operator *();
.br
.B	Type operator ++();
.br
.B	int count();
.br
.B	private:
.br
.B	List<Type> **table;
.br
.B	List<Type> currList;
.br
.B	int currHid;
.br
.B	int tableSize;
.br
};
.br
.SH "LIBRARY SYNOPSIS"
.TP 8
.B int RPCprocessCreate (pid, host, User, command,
.B	 args)
.br
.B int *pid;
.br
.B char *host;  
.br
.B char *User;           
.br
.B char *command; 
.br
.B char **args;   
.TP 8
.B int RPC_readReady (fd)
.br
.B int fd;
.TP 8
.B int RPC_setup_socket (socket_fd, family, type)
.br
.B int *socket_fd;
.br
.B int family;
.br
.B int type;
.TP 8
.B char **RPC_make_arg_list (program, family, type,
.B			     port, flag)
.br
.B char *program;
.br
.B int family;
.br
.B int type;
.br
.B int port;
.br
.B int flag;
.TP 8
.B int RPC_undo_arg_list (argc, arg_list, machine,
.B	                   family, type, socket, flag)
.br
.B int argc;
.br
.B char **arg_list;
.br
.B char **machine;
.br
.B int *family;
.br
.B int *type;
.br
.B int *socket;
.br
.B int *flag;
.TP 8
.B int RPC_getConnect (fd)
.br
.B int fd;

.SH "LIST CLASS DESCRIPTION"
.PP
.SH "HASH TABLE DESCRIPTION"
.PP
.SH "LIBRARY DESCRIPTION"
.PP
.B int RPCprocessCreate (pid, host, User, command,
.B	 args)
.br
.B int *pid;
.br
.B char *host;  
.br
.B char *User;           
.br
.B char *command; 
.br
.B char **args;   
.br
stdout will be closed for the exec'd program.  The pid is returned in pid.
Host is the name of the host to start the program on.  Command is the name
of the command to exec.  If args is non-null, it will be used as an argument 
for the exec.  Thus, command should be the first entry in args.  If args is
null, command is passed as the argument list.
.TP 8
.B int RPC_readReady (fd)
.br
.B int fd;
.TP 8
.B int RPC_setup_socket (socket_fd, family, type)
.br
.B int *socket_fd;
.br
.B int family;
.br
.B int type;
.TP 8
.B char **RPC_make_arg_list (program, family, type,
.B			     port, flag)
.br
.B char *program;
.br
.B int family;
.br
.B int type;
.br
.B int port;
.br
.B int flag;
.TP 8
.B int RPC_undo_arg_list (argc, arg_list, machine,
.B	                   family, type, socket, flag)
.br
.B int argc;
.br
.B char **arg_list;
.br
.B char **machine;
.br
.B int *family;
.br
.B int *type;
.br
.B int *socket;
.br
.B int *flag;
.TP 8
.B int RPC_getConnect (fd)
.br
.B int fd;

.SH "SEE ALSO"
Paradyn(1).
.SH COPYRIGHT
Copyright 1994, The Paradyn Software Team.
.SH AUTHOR
Mark Callaghan     (markc@cs.wisc.edu)





CHANGELOG
6890a8e Fallthrough edges should not be tail calls, ever--they should be shared code.
b2b9bed Updates for 8.1.2: version & date in manual frontpage, function definition in parseAPI manual, newly generated parseAPI manual PDF
5f8c0af Ensure that we compile test suite and Dyninst with the same CXXFLAGS to work around gcc 4.7.2 bug
93415ab Track whether we've detected that attach will break, and skip Dyninst attach mode if so.
11ace21 Prevent us accidentally rewriting indirect symbols as section typed
be7aa6b Fix a problem rewriting 32-bit static Ubuntu binaries; add better error handling.
886d8dc Fix a #define bug when instrumenting ia32 binaries
0b07ddd Use Matt's patching mechanism for while loop generation as well as if statement generation
1df2a7e Fix PPC64 proccontrol library injection; TOC value could be wrong
e30438e Handle compressed OPD sections in PPC64 binaries
3979d27 Add test for bug reported by Jean-Paul Laurent: mov %eax, 0x3(%rsp) was reporting incorrect registers.
b0add71 clean up ProcControlAPI Windows Build Directory
536628f Symlink path fix as suggested by Josh Stone


common/src/rpcUtil.C
/*
 * $Log: rpcUtil.C,v $
 * Revision 1.30  1994/09/20 18:23:58  hollings
 * added option to use rexec as well as fork and rsh to start processes.
 *
 * Revision 1.29  1994/08/18  19:55:04  markc
 * Added ifdef for solaris.
 *
 * Revision 1.28  1994/08/17  18:25:25  markc
 * Added RPCgetArg
 * Change RPC_make_arg_list to avoid leaving a hole at the head of the arg list
 * Changed RPCProcessCreate to use the new version of arg list
 * Changed the execl to execlp
 *
 * Revision 1.27  1994/07/28  22:22:04  krisna
 * changed definitions of ReadFunc and WriteFunc to conform to prototypes
 *
 * Revision 1.26  1994/07/19  18:30:27  markc
 * Made machineName default to zero as last parameter to RPC_make_arg_list.
 * Added typecast to malloc call in RPC_make_arg_list.
 *
 * Added typecast to malloc call in RPC_make_arg_list.
 *
 * Revision 1.25  1994/07/18  19:08:25  hollings
 * added extra arg to RPC_make_arg_list.
 *
 * Revision 1.24  1994/06/22  00:37:13  markc
 * Fixed code to remove warnings.
 *
 * Revision 1.23  1994/06/02  23:36:58  markc
 * Added support for igen error checking.
 *
 * Revision 1.22  1994/05/17  00:14:45  hollings
 * added rcs log entry.
 *
 * Revision 1.21  1994/05/16  04:27:47  hollings
 * Added inlcude of vfork.h on SUNS to prevent problem with optimizer.
 *
 * Revision 1.20  1994/05/12  18:47:51  jcargill
 * Changed make args function to leave room for program name in arg_list[0],
 * and added code to RPCprocessCreate to poke it in there before execv'ing.
 *
 * Revision 1.19  1994/04/21  23:23:49  hollings
 * removed paradynd name from make args function.
 *
 * Revision 1.18  1994/04/06  22:46:12  markc
 * Fixed bug in XDRrpc constructor that clobbered the fd value.  Added feature
 * to RPC_readReady to do blocking select.
 *
 * Revision 1.17  1994/04/01  20:05:27  hollings
 * Removed kill of rsh process (not needed and it causes a race condition).
 *
 * Revision 1.16  1994/04/01  04:59:13  markc
 * Put in support to encode NULL ptrs to strings in xdr_String.
 *
 * Revision 1.15  1994/03/31  22:59:08  hollings
 * added well known port as a paramter to xdrRPC constructor.
 *
 * Revision 1.14  1994/03/31  22:45:04  markc
 * Added Log for rcs.
 *
*/

//
// This file defines a set of utility routines for RPC services.
//
//

// overcome malloc redefinition due to /usr/include/rpc/types.h declaring 
// malloc 

#include <signal.h>
#include <sys/wait.h>

/* prevents malloc from being redefined */
#ifdef MIPS
#define MALLOC_DEFINED_AS_VOID
#endif

#include "util/h/rpcUtil.h"
#include "util/h/tunableConst.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <assert.h>
#include <fcntl.h>
#include <memory.h>
#include <netinet/in.h>
#include <netdb.h>
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <sys/file.h>

extern "C" {
#include <rpc/types.h>
#include <rpc/xdr.h>
}

// functions that g++-fixincludes missed
#ifdef MIPS
extern "C" {
void bzero (char*, int);
int select (int, fd_set*, fd_set*, fd_set*, struct timeval*);
char *strdup (char*);
int gethostname(char*, int);
int socket(int, int, int);
int bind(int s, struct sockaddr *, int);
int getsockname(int, struct sockaddr*, int *);
int listen(int, int);
int connect(int s, struct sockaddr*, int);
int socketpair(int, int, int, int sv[2]);
int vfork();
int accept(int, struct sockaddr *addr, int *); 
}
#elif SPARC
#include <sys/socket.h>
#include <vfork.h>
extern "C" {
void bzero (char*, int);
int select (int, fd_set*, fd_set*, fd_set*, struct timeval*);
int socket(int, int, int);
int gethostname(char*, int);
int bind(int s, struct sockaddr *, int);
int getsockname(int, struct sockaddr*, int *);
int listen(int, int);
int connect(int s, struct sockaddr*, int);
int socketpair(int, int, int, int sv[2]);
int accept(int, struct sockaddr *addr, int *); 
}
#elif SOLARIS
#endif


#define RSH_COMMAND	"rsh"

tunableBooleanConstant useRexec(FALSE, NULL, userConstant, "useRexec",
    "Use rsedc instead of rsh to establish connection to daemon");

int RPCdefaultXDRRead(const void* handle, char *buf, u_int len)
{
    int fd = (int) handle;
    int ret;

    do {
	ret = read(fd, buf, len);
    } while (ret < 0 && errno == EINTR);

    if (ret <= 0) return(-1);
    return (ret);
}

int RPCdefaultXDRWrite(const void* handle, char *buf, u_int len)
{
    int fd = (int) handle;
    int ret;

    do {
	ret = write(fd, buf, len);
    } while (ret < 0 && errno == EINTR);

    if (ret != len) 
	return(-1);
    else
	return (ret);
}

XDRrpc::~XDRrpc()
{
  if (fd >= 0)
    {
      fcntl (fd, F_SETFL, FNDELAY);
      close(fd);
    }
  if (__xdrs__) 
    {
      xdr_destroy (__xdrs__);
      delete (__xdrs__);
    }
}

//
// prepare for RPC's to be done/received on the passed fd.
//
XDRrpc::XDRrpc(int f, xdrIOFunc readRoutine, xdrIOFunc writeRoutine, int nblock)
{
    fd = f;
    __xdrs__ = new XDR;
    if (!readRoutine) readRoutine = (xdrIOFunc) RPCdefaultXDRRead;
    if (!writeRoutine) writeRoutine = (xdrIOFunc) RPCdefaultXDRWrite;
    (void) xdrrec_create(__xdrs__, 0, 0, (char *) fd, readRoutine,writeRoutine);
    if (nblock)
      fcntl (fd, F_SETFL, FNDELAY);
}

//
// prepare for RPC's to be done/received on the passed fd.
//
XDRrpc::XDRrpc(char *machine,
	       char *user,
	       char *program,
	       xdrIOFunc readRoutine, 
	       xdrIOFunc writeRoutine,
	       char **arg_list,
	       int nblock,
	       int wellKnownPortFd)
{
    fd = RPCprocessCreate(&pid, machine, user, program, arg_list, 
	wellKnownPortFd);
    if (fd >= 0) {
        __xdrs__ = new XDR;
	if (!readRoutine) readRoutine = (xdrIOFunc) RPCdefaultXDRRead;
	if (!writeRoutine) writeRoutine = (xdrIOFunc) RPCdefaultXDRWrite;
	(void) xdrrec_create(__xdrs__, 0, 0, (char *) fd, 
		readRoutine, writeRoutine);
	if (nblock)
	  fcntl (fd, F_SETFL, FNDELAY);
    } else {
	__xdrs__ = NULL;
	fd = -1;
    }
}

int
RPC_readReady (int fd, int timeout)
{
  fd_set readfds;
  struct timeval tvptr, *the_tv;

  tvptr.tv_sec = timeout; tvptr.tv_usec = 0;
  if (fd < 0) return -1;
  FD_ZERO(&readfds);
  FD_SET (fd, &readfds);

  // -1 timeout = blocking select
  if (timeout == -1)
     the_tv = 0;
  else
     the_tv = &tvptr;
 
  if (select (fd+1, &readfds, NULL, NULL, the_tv) == -1)
    {
      // if (errno == EBADF)
	return -1;
    }
  return (FD_ISSET (fd, &readfds));
}

int 
RPC_undo_arg_list (int argc, char **arg_list, char **machine, int &family,
		   int &type, int &well_known_socket, int &flag)
{
  int loop;
  char *ptr;
  int sum = 0;

  for (loop=0; loop < argc; ++loop)
    {
      if (!strncmp(arg_list[loop], "-p", 2))
	{
	  well_known_socket = (int) strtol (arg_list[loop] + 2, &ptr, 10);
	  if (ptr == (arg_list[loop] + 2))
	    return(-1);
	  sum |= 1;
	}
      else if (!strncmp(arg_list[loop], "-f", 2))
	{
	  family = (int) strtol (arg_list[loop] + 2, &ptr, 10);
	  if (ptr == (arg_list[loop] + 2))
	    return(-1);
          sum |= 2;
	}
      else if (!strncmp(arg_list[loop], "-t", 2))
	{
	  type = (int) strtol (arg_list[loop] + 2, &ptr, 10);
	  if (ptr == (arg_list[loop] + 2))
	    return(-1);
          sum |= 4;
	}
      else if (!strncmp(arg_list[loop], "-m", 2))
	{
	  *machine = strdup (arg_list[loop] + 2);
	  if (!(*machine)) return -1;
	  sum |= 8;
	}
      else if (!strncmp(arg_list[loop], "-l", 2))
	{
	  flag = (int) strtol (arg_list[loop] + 2, &ptr, 10);
	  if (ptr == (arg_list[loop] + 2))
	    return(-1);
          sum |= 16;
	}
    }
  if (sum == (16 + 8 + 4 + 2 + 1))
	return 0;
  else
	return -1;
}

/*
 * Build an argument list starting at position 0.
 * Note, this arg list will be used in an exec system call
 * AND, the command name will have to be inserted at the head of the list
 * But, a NULL space will NOT be left at the head of the list
 */
char **RPC_make_arg_list(int family, int type, int well_known_socket,
		   int flag, char *machine_name)
{
  char arg_str[100];
  int arg_count = 0;
  char **arg_list;

  arg_list = new char*[8];
  sprintf(arg_str, "%s%d", "-p", well_known_socket);  
  arg_list[arg_count++] = strdup (arg_str);  // 0
  sprintf(arg_str, "%s%d", "-f", family);
  arg_list[arg_count++] = strdup (arg_str);  // 1
  sprintf(arg_str, "%s%d", "-t", type);
  arg_list[arg_count++] = strdup (arg_str);  // 2
  if (!machine_name) {
      machine_name = (char *) malloc(50);
      gethostname (machine_name, 49);
  }
  sprintf(arg_str, "%s%s", "-m", machine_name);
  arg_list[arg_count++] = strdup (arg_str); // 3
  sprintf(arg_str, "%s%d", "-l", flag);
  arg_list[arg_count++] = strdup (arg_str);  // 4
  arg_list[arg_count++] = 0;                 // 5
  return arg_list;
}

// returns fd of socket that is listened on, or -1
int
RPC_setup_socket (int *sfd,   // return file descriptor
		  int family, // AF_INET ...
		  int type)   // SOCK_STREAM ...
{
  struct sockaddr_in serv_addr;
  int length;
  char machine[50];

  if (gethostname(machine, 49) != 0)
    return -1;

  if ((*sfd = socket(family, type, 0)) < 0)
    return -1;

  memset ((char*) &serv_addr, 0, sizeof(serv_addr));
  /* bzero ((char *) &serv_addr, sizeof(servaddr)); */
  serv_addr.sin_family = (short) family;
  serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
  serv_addr.sin_port = htons(0);
  
  length = sizeof(serv_addr);

  if (bind(*sfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0)
    return -1;

  if (getsockname (*sfd, (struct sockaddr *) &serv_addr, &length) < 0)
    return -1;

  if (listen(*sfd, 5) < 0)
    return -1;

  return (ntohs (serv_addr.sin_port));
}

//
// connect to well known socket
//
XDRrpc::XDRrpc(int family,            
	       int req_port,             
	       int type,
	       char *machine, 
	       xdrIOFunc readRoutine,
	       xdrIOFunc writeRoutine,
	       int nblock)
     // socket, connect using machine
{
  struct sockaddr_in serv_addr;
  struct hostent *hostptr = 0;
  struct in_addr *inadr = 0;

  __xdrs__ = 0;

  if ( (hostptr = gethostbyname(machine)) == 0)
    { fd = -1; return; }

  inadr = (struct in_addr *) hostptr->h_addr_list[0];
  memset ((char*) &serv_addr, 0, sizeof(serv_addr));
  /* bzero ((char *) &serv_addr, sizeof(serv_addr)); */
  serv_addr.sin_family = family;
  serv_addr.sin_addr = *inadr;
  serv_addr.sin_port = htons(req_port);

  if ( (fd = socket(family, type, 0)) < 0)
    { fd = -1; return; }

  if (connect(fd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0)
    { fd = -1; return; }

    __xdrs__ = new XDR;
    if (!readRoutine) readRoutine = (xdrIOFunc) RPCdefaultXDRRead;
    if (!writeRoutine) writeRoutine = (xdrIOFunc) RPCdefaultXDRWrite;
    (void) xdrrec_create(__xdrs__, 0, 0, (char *) fd, readRoutine,writeRoutine);
    if (nblock)
      fcntl (fd, F_SETFL, FNDELAY);

}

//
// prepare for RPC's to be done/received on the passed thread id.
//
THREADrpc::THREADrpc(int thread)
{
    tid = thread;
}

//
// This should never be called, it should be replaced by a virtual function
//    from the derived class created by igen.
//
void RPCUser::verifyProtocolAndVersion()
{
    abort();
}

//
// our version of string encoding that does malloc as needed.
//
bool_t xdr_String(XDR *xdrs, String *str)
{
    int len;
    unsigned char isNull=0;

	// if XDR_FREE, str's memory is freed
    switch (xdrs->x_op) {
	case XDR_ENCODE:
		if (*str) {
		    len = strlen(*str)+1;
                    if (!xdr_u_char(xdrs, &isNull))
			return FALSE;
                } else {
		    isNull = (unsigned char) 1;
                    if (!xdr_u_char(xdrs, &isNull))
			return FALSE;
                    else
			return TRUE;
                }
                return (xdr_string (xdrs, str, 65536));
	case XDR_DECODE:
		*str = NULL;
		if (!xdr_u_char(xdrs, &isNull))
		    return FALSE;
                if (isNull)
		    return TRUE;
                else
                    return (xdr_string (xdrs, str, 65536));
	case XDR_FREE:
		// xdr_free (xdr_string, str);
		if (*str)
		  free (*str);
		*str = NULL;
		return (TRUE);
		// return(TRUE);
		// free the memory
	default:
                return (FALSE);
		// this should never occur	
    }
}

//
// directly exec the command (local).
//
int execCmd(int *pid, char *command, char **arg_list, int portFd)
{
    int ret;
    int sv[2];
    char **new_al;
    int al_len, i;
    int execlERROR;

    ret = socketpair(AF_UNIX, SOCK_STREAM, 0, sv);
    if (ret) return(ret);
    execlERROR = 0;
    *pid = vfork();
    if (*pid == 0) {
	close(sv[0]);
	dup2(sv[1], 0);
	if (!arg_list)
	  execlp(command, command);
	else {
	  // how long is the arg_list?
	  for (al_len=0; arg_list[al_len]; al_len++) 
	    ; // not a typo

	  new_al = new char*[al_len+2];
	  new_al[0] = command;
	  new_al[al_len+1] = 0;
	  for (i=0; i<al_len; ++i) {
	    new_al[i+1] = arg_list[i];
	  }
	  execvp(command, new_al);
	  free(new_al);
	}
	execlERROR = errno;
	_exit(-1);
    } else if (*pid > 0 && !execlERROR) {
	close(sv[1]);
	return(sv[0]);
    } else {
	return(-1);
    }
    return(-1);
}

int handleRemoteConnect(int *pid, int fd, int portFd)
{
    char *ret;
    int retFd;
    FILE *pfp;
    char line[256];

    pfp = fdopen(fd, "r");
    do {
	ret = fgets(line, sizeof(line)-1, pfp);
	if (ret && !strncmp(line, "PARADYND", strlen("PARADYND"))) {
	    // got the good stuff
	    sscanf(line, "PARADYND %d", pid);

	    retFd = RPC_getConnect(portFd);
	    return(retFd);
	} else if (ret) {
	    printf("%s", line);
	    return (-1);
	}
    }  while (ret);

    return(-1);
}

//
// use rsh to get a remote process started.
//
int rshCommand(int *pid, char *hostName, char *userName, char *command, 
    char **arg_list, int portFd)
{
    int total;
    int fd[2];
    char **curr;
    int shellPid;
    char *paradyndCommand;

    total = strlen(command) + 2;
    for (curr = arg_list; *curr; curr++) {
	total += strlen(*curr) + 2;
    }
    paradyndCommand = (char *) malloc(total+2);

    sprintf(paradyndCommand, "%s ", command);
    for (curr = arg_list; *curr; curr++) {
	strcat(paradyndCommand, *curr);
	strcat(paradyndCommand, " ");
    }

    // need to rsh to machine and setup io path.

    if (pipe(fd)) {
	perror("pipe");
	return (-1);
    }

    shellPid = vfork();
    if (shellPid == 0) {
	/* child */
	dup2(fd[1], 1);                         /* copy it onto stdout */
	close(fd[0]);
	close(fd[1]);
	if (userName) {
	    execlp(RSH_COMMAND, RSH_COMMAND, hostName, "-l", 
		userName, "-n", paradyndCommand, "-l0", NULL);
	} else {
	    execlp(RSH_COMMAND, RSH_COMMAND, hostName, "-n", 
		paradyndCommand, "-l0", NULL);
	}
	_exit(-1);
    } else if (shellPid > 0) {
	close(fd[1]);
    } else {
	// error situation
    }
    return(handleRemoteConnect(pid, fd[0], portFd));
}

int rexecCommand(int *pid, char *hostName, char *userName, char *command, 
    char **arg_list, int portFd)
{
    int fd;
    int total;
    char **curr;
    char *paradyndCommand;
    struct servent *inport;

    total = strlen(command) + 2;
    for (curr = arg_list; *curr; curr++) {
	total += strlen(*curr) + 2;
    }
    paradyndCommand = (char *) malloc(total+2);

    sprintf(paradyndCommand, "%s ", command);
    for (curr = arg_list; *curr; curr++) {
	strcat(paradyndCommand, *curr);
	strcat(paradyndCommand, " ");
    }

    inport = getservbyname("exec",  "tcp");

    fd = rexec(&hostName, inport->s_port, userName,NULL, paradyndCommand, NULL);
    if (fd < 0) {
	perror("rexec");
	printf("rexec failed\n");
    }
    return(handleRemoteConnect(pid, fd, portFd));
}

/*
 * arg_list should not have "command" at arg_list[0], it will
 * be put there
 *
 * if arg_list == NULL, command is the only argument used
 */
int RPCprocessCreate(int *pid, char *hostName, char *userName,
		     char *command, char **arg_list, int portFd)
{
    int ret;
    char local[50];

    if (gethostname(local, 49))
	strcpy (local, " ");

    if (!hostName || 
	!strcmp(hostName, "") || 
	!strcmp(hostName, "localhost") ||
	!strcmp(hostName, local)) {
	ret = execCmd(pid, command, arg_list, portFd);
    } else if (useRexec.getValue()) {
	ret = rexecCommand(pid, hostName, userName, command, arg_list, portFd);
    } else {
	ret = rshCommand(pid, hostName, userName, command, arg_list, portFd);
    }
    return(ret);
}

//
// wait for an expected RPC responce, but also handle upcalls while waiting.
//    Should not be called directly !!!
//
void RPCUser::awaitResponce(int tag)
{
    abort();
}

void
XDRrpc::setNonBlock()
{
  if (fd >= 0)
    fcntl (fd, F_SETFL, FNDELAY);
}

int
XDRrpc::readReady(int timeout)
{
  return RPC_readReady (fd, timeout);
}

int
RPC_getConnect(int fd)
{
  int clilen;
  struct in_addr cli_addr;
  int new_fd;

  if (fd == -1)
    return -1;

  clilen = sizeof(cli_addr);

  if ((new_fd = accept (fd, (struct sockaddr *) &cli_addr, &clilen)) < 0)
    return -1;
  else
    return new_fd;
}

RPCUser::RPCUser(int st)
{
  err_state = st;
}

RPCServer::RPCServer(int st)
{
  err_state = st;
}

/*
 *  RPCgetArg - break a string into blank separated words
 *  Used to parse a command line.
 *  
 *  input --> a null terminated string, the command line
 *  argc --> returns the number of args found
 *  returns --> words (separated by blanks on command line)
 *  Note --> this allocates memory 
 */
char **RPCgetArg(int &argc, const char *input)
{
#define BLANK ' '

  char **temp, **result;
  const char *word_start;
  int word_count = 0, temp_max, index, length, word_len;

  argc = 0;
  if (!input) 
    return ((char **) 0);

  length = strlen(input);
  index = 0;

  /* advance past blanks in input */
  while ((index < length) && (input[index] == BLANK))
    index++;

  /* input is all blanks, or NULL */
  if (index >= length) {
    result = new char*[1];
    if (!result) return ((char**) 0);
    result[0] = 0;
    return result;
  }

  temp = new char*[30];
  if (!temp) return temp;
  temp_max = 29;

  do {
    /* the start of each string */
    word_len = 0; word_start = input + index;

    /* find the next BLANK and the WORD size */
    while ((index < length) && (input[index] != BLANK)) {
      index++; word_len++;
    }

    /* copy the word */
    temp[word_count] = new char[word_len+1];
    strncpy(temp[word_count], word_start, word_len);
    temp[word_count][word_len] = (char) 0;
    word_count++;

    /* skip past consecutive blanks */
    while ((index < length) && (input[index] == BLANK))
      index++;

    /* no more room in temp, copy it to new_temp */
    if (word_count > temp_max) {
      char **new_temp;
      int nt_len, i;
      
      /* new temp_max size */
      nt_len = (temp_max+1) >> 1;
      temp_max = nt_len - 1;

      /* copy temp to new_temp */
      new_temp = new char*[nt_len];
      if (!new_temp) return ((char**) 0);
      for (i=0; i<word_count; ++i)
	new_temp[i] = temp[i];
      
      delete temp;
      temp = new_temp;
    }
  } while (index < length);

  argc = word_count;
  /* null terminate the word list */
  temp[word_count] = (char*) 0;
  return temp;
}
